text/0249-v2-experiments.md

```diff
@@ -0,0 +1,127 @@
+  the v2 branch, and moved to the preview repository. This option should be minimized, since some customers may have
+  deployed production stacks that depend on these APIs.
+- Create a new repository called aws-cdk-previews that will be the new home for experimental development and developer
+  preview releases. The packages in this repository will be vended in the same way that 1.x packages are vended today,
+  in a variety of languages, with metadata that draws a line between new experiments and more mature developer preview
+  packages. This repo will take a dependency on the core CDK package just like any other independent library. This
+  avoids the complications associated with experimenting on a fork or a branch of the main repository. The repository
+  will be clearly marked as experimental.
+- When a package in the preview repository has reached maturity and is ready for GA, it will be copied into the main
+  repo and made a part of CDK 2. The module in the preview repository will then be marked as deprecated with clear
+  documentation to point users to the new stable version in the main repository. The lifecycle of experimental modules
+  in the preview repository will be exactly the same as it is currently in 1.x, with the same metadata markers in the
+  code.
+- The contributor guide for the CDK will be modified so that users know where to introduce new experimental code.
+- If it becomes absolutely necessary to add a new feature to v2 that would be a breaking change for users, the feature
+  will be hidden behind a feature flag. The code path will not be accessible without a specific opt-in by the user.
+
+# Drawbacks
+
+- In several places in the current 1.x CDK branch, stable modules refer to or contain experimental code. This will no
+  longer be possible after the release of v2.
+- When we graduate a module, users who have participated in the Developer Preview will have to update their imports to
+  make use of the stable version in the main repository.
+- For developers who are working on experimental code that has complex dependencies with the core and with other
+  modules, development will be more complicated.
+- In some cases, experimentation will not be possible. New features introduced to the main repository will have to be
+  vetted much more carefully before they are merged, since any public API released as part of v2 will have to be
+  supported for the lifetime of the major version.
+
+# Alternatives
+
+Some alternatives to this strategy are:
+
+- Make no changes and continue to experiment in the main release. Improve documentation so that it is more obvious to
+  users that they are importing experimental code. We might also make it a policy to actually include the name
+  'experimental' in all experimental APIs.
+- Use clever build tricks to hide experimental modules. This might be possible in Typescript, where the type files are
```

Comment:
Here's another alternative: use runtime checks.

Calling an experimental API throws an exception which can be silenced by flipping a flag `ALLOW_EXPERIMENTAL=1`.

Reply:
Adding that to the next commit.

Reply:
I think this is my preferred approach for enabling experimentation within stable modules. It requires some form of enforcement (awslint/jsii) which will make sure that people always include some runtime check when an API is annotated as @experimental (or jsii injecting some code somehow).

Reply:
> (or jsii injecting some code somehow)

I'm not entirely against the idea of adding a `jsii` compiler flag to trigger the injection...  I'm just wondering if this is a good fit for it (should we consider this is a "language primitive"?

Frankly, I'm thinking it might be more appropriate then to swap out the `@experimental` TS-Doc tag with an `@experimental` decorator (however decorators are experimental themselves.....)

Reply:
Yes I think decorators will the superior solution

Reply:
Decorators would definitely be a great solution, but what are the chances of things changing before it's a stable feature? Seems like too much of a risk for now.